core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultAwsV4HttpSigner.java

```diff
@@ -15,27 +15,228 @@
+
+        return requestBuilder;
+    }
+
+
+    private void addHostHeader(SdkHttpRequest.Builder requestBuilder) {
+        // AWS4 requires that we sign the Host header, so we
+        // have to have it in the request by the time we sign.
+
+        StringBuilder hostHeaderBuilder = new StringBuilder(requestBuilder.host());
+        if (!SdkHttpUtils.isUsingStandardPort(requestBuilder.protocol(), requestBuilder.port())) {
+            hostHeaderBuilder.append(":").append(requestBuilder.port());
+        }
+
+        requestBuilder.putHeader(SignerConstant.HOST, hostHeaderBuilder.toString());
+    }
+
+    private void addDateHeader(SdkHttpRequest.Builder requestBuilder, String dateTime) {
+        requestBuilder.putHeader(SignerConstant.X_AMZ_DATE, dateTime);
+    }
+
+    private void putChecksumHeader(SdkChecksum sdkChecksum,
+                                   SdkHttpRequest.Builder requestBuilder, String contentHashString, String headerChecksum) {
+
+        if (sdkChecksum != null && !UNSIGNED_PAYLOAD.equals(contentHashString)
+            && !"STREAMING-UNSIGNED-PAYLOAD-TRAILER".equals(contentHashString)) {
+
+            if (HttpChecksumUtils.isHttpChecksumPresent(requestBuilder.build(),
+                ChecksumSpecs.builder()
+                    .headerName(headerChecksum).build())) {
+                LOG.debug(() -> "Checksum already added in header ");
+                return;
+            }
+            if (StringUtils.isNotBlank(headerChecksum)) {
+                requestBuilder.putHeader(headerChecksum,
+                    BinaryUtils.toBase64(sdkChecksum.getChecksumBytes()));
+            }
+        }
     }
 
     @Override
     public AsyncSignedHttpRequest signAsync(AsyncHttpSignRequest<? extends AwsCredentialsIdentity> request) {
-        throw new UnsupportedOperationException();
+        if (CredentialUtils.isAnonymous(request.identity())) {
+            return AsyncSignedHttpRequest.builder()
+                .request(request.request())
+                .payload(request.payload().orElse(null))
+                .build();
+        }
+
+        SdkChecksum sdkChecksum = createSdkChecksumFromRequest(request.request(), request.property(CHECKSUM_HEADER_NAME),
+            request.property(CHECKSUM_ALGORITHM));
+        DigestComputingSubscriber bodyDigester = DigestComputingSubscriber.forSha256(sdkChecksum);
+
+        request.payload().ifPresent((payload) ->
+            payload.subscribe(bodyDigester)
+        );
+
+        CompletableFuture<byte[]> digestBytes = bodyDigester.digestBytes();
+
+        CompletableFuture<SdkHttpRequest> signedReqFuture = digestBytes.thenApply(bodyHash -> {
+            String digestHex = BinaryUtils.toHex(bodyHash);
+
+            SdkHttpRequest.Builder builder = doSign(request,
+                new ContentChecksum(digestHex, sdkChecksum));
+
+            return builder.build();
+        });
+
+        return AsyncSignedHttpRequest.builder()
+            .request(signedReqFuture.join())
```

Comment:
We should use `CompletableFutureUtils.joinLikeSync`.

@gosar question: sorry if I missed the previous conversion, but did we decide to block on signing?

Reply:
Hmm, good question! I don't think I at least had a specific conversation about it. The SRA and Java-specific design docs didn't use CompletableFuture for the sign response in the interfaces (as opposed to the Identity interfaces that do), so I assumed this was vetted in original SRA design. But maybe it's a miss. @millems can you weigh in?

For the purpose of this PR though, I'd say maybe do it as per the current interface (using CompletableFutureUtils.joinLikeSync), and do another PR to update the interfaces if we decide to.